src/mpid/ch4/shm/xpmem/xpmem_avl.h

```diff
@@ -0,0 +1,564 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef XPMEM_AVL_H_INCLUDED
+#define XPMEM_AVL_H_INCLUDED
+
+#include "mpidimpl.h"
+#include "xpmem_pre.h"
+#include <xpmem.h>
+/*
```

Comment:
Why do you need `mpidimpl.h` ?

Once you removed `xpmem_detach` in the file, you do not need `<xpmem.h>` too.

Reply:
I need a header file to provide `MPIR_FUNC_VERBOSE_STATE_DECL`, `MPIR_FUNC_VERBOSE_ENTER` and `MPIR_FUNC_VERBOSE_EXIT` definitions. Before I add `xpmem_pre.h`, if I don't add `mpidimpl.h`, it will report compilation error. I think `xpmem_pre.h` now includes the upper layer header file (`mpi.h`), so I will delete it anyway. 
However, I am confused about the header file including. Should I include the necessary header file from the upper layer? I feel upper layer header file has include every thing from the same layer and below. Is there a document that talks about how header file is organized?

Reply:
We do not have good documentation for it. Just delete unnecessary ones if you confirmed it works :-)